src/Hypsography/Hypsography.jl

```diff
@@ -64,180 +119,154 @@ If the decay-scale is poorly specified (i.e., `s * zₜ` is lower than the maxim
 )
-    ηₕ = adaption.ηₕ
-    s = adaption.s
-    FT = eltype(s)
-    @assert FT(0) <= ηₕ <= FT(1)
-    @assert s >= FT(0)
-    if s * z_top <= maximum(z_surface)
-        @warn "Decay scale (s*z_top = $(s*z_top)) must be higher than max surface elevation (max(z_surface) = $(maximum(z_surface))). Returning s = FT(0.8). Scale height is therefore s=$(0.8 * z_top) m."
+    (; ηₕ, s) = adaption
+    if s * z_top.z <= z_surface.z
+        error("Decay scale (s*z_top) must be higher than max surface elevation")
     end
-end
 
-function ref_z_to_physical_z(adaption::SLEVEAdaption, z_ref, z_surface, z_top)
-    ηₕ = adaption.ηₕ
-    s = adaption.s
-    η = z_ref ./ z_top
-    FT = eltype(s)
-    # Fix scale height if needed
-    s = ifelse(s * z_top <= maximum(z_surface), FT(8 / 10), s)
-    return ifelse(
-        η <= ηₕ,
-        η * z_top + z_surface * (sinh((ηₕ - η) / s / ηₕ)) / (sinh(1 / s)),
-        η * z_top,
-    )
+    η = z_ref.z / z_top.z
+    if η <= ηₕ
+        return Geometry.ZPoint(
+            η * z_top.z +
+            z_surface.z * (sinh((ηₕ - η) / s / ηₕ)) / (sinh(1 / s)),
+        )
+    else
+        return Geometry.ZPoint(η * z_top.z)
+    end
 end
 
-# deprecated in 0.10.12
-LinearAdaption() = LinearAdaption(nothing)
-
-@deprecate(
-    ExtrudedFiniteDifferenceSpace(
-        horizontal_space::Spaces.AbstractSpace,
-        vertical_space::Spaces.FiniteDifferenceSpace,
-        adaption::LinearAdaption{Nothing},
-        z_surface::Fields.Field,
-    ),
-    ExtrudedFiniteDifferenceSpace(
-        horizontal_space,
-        vertical_space,
-        LinearAdaption(z_surface),
-    ),
-    false
+function physical_z_to_ref_z(
+    adaption::SLEVEAdaption,
+    z_physical::Geometry.ZPoint,
+    z_surface::Geometry.ZPoint,
+    z_top::Geometry.ZPoint,
 )
+    error("This method is not implemented")
+end
 
-# linear coordinates
+# can redefine this constructor for e.g. multi-arg SLEVE
 function _ExtrudedFiniteDifferenceGrid(
     horizontal_grid::Grids.AbstractGrid,
     vertical_grid::Grids.FiniteDifferenceGrid,
     adaption::HypsographyAdaption,
+    global_geometry::Geometry.AbstractGlobalGeometry,
 )
-    if adaption isa LinearAdaption
-        if isnothing(adaption.surface)
-            error("LinearAdaption requires a Field argument")
-        end
-        if axes(adaption.surface).grid !== horizontal_grid
-            error("Terrain must be defined on the horizontal space")
-        end
-    end
+    @assert Spaces.grid(axes(adaption.surface)) == horizontal_grid
+    z_surface = Fields.field_values(adaption.surface)
+
+    face_z_ref =
+        Grids.local_geometry_data(vertical_grid, Grids.CellFace()).coordinates
+    vertical_domain = Topologies.domain(vertical_grid)
+    z_top = vertical_domain.coord_max
 
-    # construct initial flat space, then mutate
-    ref_grid = Grids.ExtrudedFiniteDifferenceGrid(
+    face_z =
+        ref_z_to_physical_z.(Ref(adaption), face_z_ref, z_surface, Ref(z_top))
```

Comment:
Is `Ref` necessary here ?

Reply:
I don't know, we can check.

Reply:
Adapting the `z_surface` to `CuArray`, I think we can use `adaption` and `z_top::ZPoint` directly; then it seems the types correspond to those expected by the `_ExtrudedFiniteDifferenceGrid` constructor.